internal/internal_workflow_testsuite.go

```diff
@@ -2910,10 +2924,32 @@ func (env *testWorkflowEnvironmentImpl) updateWorkflow(name string, id string, u
 	if err != nil {
 		panic(err)
 	}
-	env.postCallback(func() {
-		// Do not send any headers on test invocations
-		env.updateHandler(name, id, data, nil, uc)
-	}, true)
+
+	if env.updateMap == nil {
+		env.updateMap = make(map[string]updateResult)
+	}
+
+	var ucWrapper = updateCallbacksWrapper{uc: uc, env: env, updateID: id}
+
+	// check for duplicate update ID
+	if result, ok := env.updateMap[id]; ok {
+		result.mu.Lock()
```

Comment:
This doesn't really address the problem because now your `updateWorkflow` call will block which will block other callbacks. To properly dedup you would need to keep track of the inflight updates requests and attach any other requests to them.

Given that is very complex I think it would be simpler to just fail if we detect a duplicate update in the test environment because there is no reason a user should ever send one.

Reply:
from a high level, could inflight updates be as easy as `updateResults` being
```
 updateResult struct {
		success interface{}
		err     error
		callbacks []func()
	}
```
and in the callback of the first request, we `defer` with a `.complete_callbacks` to `postCallback` all pending duplicate updates?

I'm gonna try spending ~20 minutes this afternoon seeing if this can work, but lmk if anything sounds wrong here at a high level. And if I can't get it to work, I'll go the "fail duplicate updates` route

Reply:
I think I got a working solution, but the only downside is maybe the `TestWorkflowDuplicateIDDedupInterweave` test isn't obvious (maybe not guaranteed?) that this exact order of code is happening. 

I've confirmed through debug print logs that 
1. both updates get scheduled
2. caches the result
3. first update completes
4. 2nd update then finishes with cached result 
 
<img width="672" alt="Screenshot 2024-11-21 at 3 14 59 PM" src="https://github.com/user-attachments/assets/b1750c64-2690-4551-9c9a-0fa1b18bfff5">

not sure if there's a more explicit way of showing/guaranteeing that this exact order is happening.